README.md

```diff
@@ -2,7 +2,217 @@
 
 TASTy Query will be the basis for all compiler-independent tools that analyze TASTy - an intermediate representation of Scala 3 code.
 
-It scans the classpath to build a map of all definitions in a project. Its API will allow users to query the code for a class, inspect signatures of methods, values and types, and compare them.
+It scans the classpath to build a map of all definitions in a project.
+Its API will allow users to query the code for a class, inspect signatures of methods, values and types, and compare them.
+
+TASTy Query is still in development.
+It is possible to obtain cursory information about trees, types and symbols in TASTy files, but deep semantic information like subtyping are not there yet.
+
+## Motivation
+
+### Preamble: TASTy and separate compilation
+
+When compiling a codebase, a Scala compiler combines the source files of a project with the "binaries" of its dependencies.
+Since the project source files can reference symbols (all kinds of definitions in a Scala program, like `val`s and `class`es) from the dependencies, the compiler builds a representation of the dependencies.
+
+In Scala 2, this representation is limited to public definitions and their API-level types (also protected and package-private, but for the intent of this explanation, we refer to all of that as "public").
+This information is stored in Scala 2 class files as "pickles".
+The bodies of methods are excluded from this treatment.
+
+In Scala 3, virtually everything about the source program is stored in so-called TASTy files: the complete AST, with all the bodies, and the types of everything.
+Or to be precise, enough information is stored to be able to *reconstruct* everything (in particular, the types of subexpressions in method bodies).
+
+This representation is used to type-check (and further "elaborate") the project source files, clearly.
+In Scala 3, it is also used to process `inline def`s, which is why the full trees for method bodies are required.
+
+### Definitions
+
+* Symbol: the identity of any definition in a Scala or Java program.
+  Examples include class definitions, local or member val/def definitions.
+  There is only symbol per *definition*: multiple references to the same local val all refer to the same symbol.
+* Tree: an Abstract Syntax Tree (AST) or a subtree thereof.
+  In the context of tasty-query, they refer to TASTy trees, which carry their symbol and type information.
+* Type: a type in a Scala program, such as `Int`, `List[String]` or `Bar { type Foo <: Baz[String] }`.
+* Pickle: the public symbol and type information stored in Scala 2 binaries.
+* Elaboration: the process of turning a Scala source program into a typed AST with TASTy-level information.
+  Elaboration includes type inference, implicit resolution, overload resolution and macro expansion.
```

Comment:
```suggestion
  Elaboration includes type inference, context arguments resolution, overload resolution and macro expansion.
```

Reply:
AFAICT, the *process* of resolving these things, including contextual abstractions, is still called "implicit resolution". At least, that's how the compiler calls it.